### PR TITLE
fix transpose optimizer on GPU EP

### PR DESCRIPTION
**Pull request by fs-eire**
### Description
because of #15618 , the default allocator changed to device allocator, which will be GPU instead of CPU. in transpose optimizer we expect to read data from initializers so a CPU allocator is required here.

this change fixes transpose optimizer on GPU EP

Fixes the issue referred to in #15869, #15796

### CODE REVIEW

**Review by guschmue — DISMISSED** [review]

**Comment by askhade** [comment]
Looks like this PR need to be cherry picked as well.

File: onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc [hunk 1]
```diff
@@ -913,7 +913,15 @@ PostLayoutTransformCostCheck(const api::GraphRef& graph, const api::NodeRef& nod
 Status TransformLayoutForEP(Graph& graph, bool& modified, const IExecutionProvider& execution_provider,
                             const DebugGraphFn& debug_graph_fn) {
   // sub graph recurse will be added later
-  auto api_graph = MakeApiGraph(graph, execution_provider.GetAllocator(OrtMemTypeDefault), nullptr);
+  auto cpu_allocator = execution_provider.GetAllocator(OrtMemTypeDefault);
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc:916]: Would it be simpler for the caller to pass in the value returned by ExecutionProviders.GetDefaultCpuAllocator()?

Comment (fs-eire) [reply onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc:916]: it's used in this way -
```
Status PartitionOrtFormatModel(onnxruntime::Graph& graph,
                               const ExecutionProviders& providers,
                               KernelRegistryManager& kernel_registry_manager,
                               SessionState& session_state) {

  layout_transformer::TransformLayoutFunction transform_layout_fn = layout_transformer::IsSupportedOpset(graph)
                                                                        ? layout_transformer::TransformLayoutForEP
                                                                        : nullptr;

  GraphPartitioner partitioner(kernel_registry_manager, providers);
  ORT_RETURN_IF_ERROR(partitioner.Partition(graph,
                                            session_state.GetMutableFuncMgr(),
                                            transform_layout_fn,
                                            GraphPartitioner::Mode::kOrtFormatLoad));

  return Status::OK();
}

```

So this is going to modify the function signature of `GraphPartitioner::Partition()` and some callback type definition, which is not a trivial one.

Comment (skottmckay) [reply onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc:916]: In InferenceSessions.cs there is a lambda used so technically you'd only need to update TransformLayoutForEP and not the entire path through GraphPartitioner. 

https://github.com/microsoft/onnxruntime/blob/648bedf91a9e390fadbbabdbbb9d5239f5a9d201/onnxruntime/core/session/inference_session.cc#L944

Comment (fs-eire) [reply onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc:916]: There are 2 places where `TransformLayoutForEP()` is called, and neither of them has a context with an existing allocator. If I make this change, I still need code to get allocator from the EP, and I need do this twice.

**Review by skottmckay — DISMISSED** [review]

File: onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc (hunk 1 above)
Comment (fs-eire) [reply onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc]: Updated the code

Comment (fs-eire) [reply onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc]: Is there an easier way to create lambdas in [session_state_test.cc](https://github.com/microsoft/onnxruntime/pull/15988/files#diff-2e498c0ec894ae05b0ddfb15f7bc4b927df6626c0497c47de85a5e808317d3f1)? is there a way to do currying in a simpler way

**Review by skottmckay — APPROVED** [review]

**Review by guschmue — APPROVED** [review]

**Pull request merged**